8261160: Add a deserialization JFR event

ChrisHegarty opened the pull request:
This issue adds a new event to improve diagnostic information of Java deserialization. The event captures the details of deserialization activity from ObjectInputStream. The event details are similar to that of the serial filter, but is agnostic of whether a filter is installed or not. The event also captures the filter status, if there is one.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261160](https://bugs.openjdk.java.net/browse/JDK-8261160): Add a deserialization JFR event


### Reviewers
 * [Sean Coffey](https://openjdk.java.net/census#coffeys) (@coffeys - **Reviewer**)
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**)
 * [Erik Gahlin](https://openjdk.java.net/census#egahlin) (@egahlin - **Reviewer**)


### Contributors
 * Sean Coffey `<coffeys@openjdk.org>`
 * Chris Hegarty `<chegar@openjdk.org>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2479/head:pull/2479`
`$ git checkout pull/2479`

ChrisHegarty commented:
/help

ChrisHegarty commented:
/contributor add @coffeys

ChrisHegarty commented:
/contributor add @ChrisHegarty

ChrisHegarty commented:
The logging in ObjectInputStream remains conditional on whether a filter is installed, which that seems reasonable since the logging is filter specific, while the JRF event is not (but both carry effectively the same information).

The new jdk.Deserialization event uses a String to carry the filterStatus value. The value could be represented by its enum ordinal, but then the tool consuming the event would need to map that back to its string value to be meaningful.

coffeys commented on test/jdk/java/io/Serializable/serialFilter/GlobalFilterTest.java:
```diff
@@ -48,6 +48,7 @@
  *          -Dexpected-jdk.serialFilter=java.** GlobalFilterTest
  * @run testng/othervm/policy=security.policy GlobalFilterTest
  * @run testng/othervm/policy=security.policy
```
You may want to add a "@requires vm.hasJFR" condition to this test

dfuch commented on src/jdk.jfr/share/classes/jdk/jfr/events/DeserializationEvent.java:
```diff
@@ -0,0 +1,61 @@
+import jdk.jfr.internal.MirrorEvent;
+
+@Category({"Java Development Kit", "Serialization"})
+@Label("Deserialization events")
```
This seems to differ from the format of other event labels defined in this package, e.g:

```
@Label("Process Start")
@Label("File Read")
...
```

What would you think of changing it to one of:

```
 @Label("Deserialization")
 @Label("Deserialized Object")
```

RogerRiggs commented:
As proposed, events are only created if there is a serialFilter in effect (and enabled by JFR configuration).
Being able to create the events without a serialFilter in effect would be useful for monitoring, especially if it could be controlled by a separate JFR configuration option.  (always, never, serial-filter , etc.)

ChrisHegarty commented:
I updated the PR and addressed all comments so far. Specifically:

@RogerRiggs The generation of the event is independent of whether the filter is set or not.  I also added a piece of state to determine if a filter is set or not. I think it could be useful to analyse all Deserialisation events to, say, ensure that there are none operating without a filter ( and the `filterStatus` state is ambitious in the `null` case ).

@coffeys I would like GlobalFilterTest to run regardless of whether or not the jfr module is present, but of course running the test with jfr enabled is desirable too, so I added a separate at test tag for that case.

@egahlin Excellent suggestions on the naming, etc. I adopted all.  And added a test to ensure that the creation and generation of the event does not inadvertently trigger class initialization if the filter rejects the attempt ( thanks @dfuch )

@dfuch Thanks for the improved label suggestion, it is now merged in.

dfuch commented on src/jdk.jfr/share/classes/jdk/jfr/events/DeserializationEvent.java:
```diff
@@ -32,30 +32,33 @@
 @MirrorEvent(className = "jdk.internal.event.DeserializationEvent")
 public final class DeserializationEvent extends AbstractJDKEvent {
 
+    @Label("Filter configured")
+    public boolean filterConfigured;
```
Should that be "Filter Configured" for consistency?

ChrisHegarty replied:
Good catch. Fixed.

dfuch reviewed (approved):
LGTM

ChrisHegarty commented:
Speaking to Erik offline, he suggested to split the `exceptionMessage` into `exceptionType` and `exceptionMessage` ( since the code was somewhat overloading the existing Sting field by using toString to force the exception class name and message into a single sting ). While the exceptionXXX field should rarely be non-null, they add very little overhead.

ChrisHegarty commented:
/integrate
